Partial update

LoganDark opened the pull request:
This updates all the dependencies to the latest versions, partially. ~~Lots of features are not yet working and attempting to run any of the examples causes `thread 'main' panicked at 'gl function was not loaded'`.~~

~~Event polling is not really a thing anymore so I was thinking of having the user run the event loop themselves and pipe events to `mini_gl_fb` along the way. Of course `glutin_handle_basic_input` could probably run the event loop itself.~~

I know this library hasn't been updated in years but it's probably the only good pixel library that can be used with winit.

LoganDark committed:
Partial update

LoganDark committed:
Remove redundant EventLoop return

LoganDark commented:
I've confirmed a few things:

- `context.get_proc_address` is getting called with `glCreateShader` and is returning a valid pointer.
- However, this pointer is somehow not being assigned to `gl::storage::CreateShader`, even though `gl::CreateShader::load_with` is being called (calling it manually does nothing)
- The function pointer remains at the "gl function was not loaded" method, which panics

I guess I'm going to have to futz with my debugger until I can figure out why this is happening.

LoganDark commented:
- The returned pointer is DEFINITELY assigned to `storage::CreateShader`
  
  ![image](https://user-images.githubusercontent.com/4723091/104844303-6d714800-5884-11eb-98bd-9ed9136d809e.png)
  
  This is different from the address of `missing_fn_panic`:
  
  ![image](https://user-images.githubusercontent.com/4723091/104844319-837f0880-5884-11eb-99a9-96400d036547.png)

- `gl::raw::create_shader` simply **doesn't care and calls `missing_fn_panic` anyway**

LoganDark committed:
gl 0.14.0 is cursed

LoganDark commented:
I've found my issue!

`gl` 0.14.0 is cursed. Switching back to 0.10.0 fixed all my problems.

@Lokathor recommended moving to gl33 or gl46 as an alternative. That may be a separate PR in the future.

LoganDark committed:
Fix event loop code, basic input, game of life example

LoganDark commented:
Game of life is now working perfectly, but custom shader is panicking and breakout example is out of date

LoganDark committed:
Refactor all code that handles event loops

You should not keep this internal and hide it from the user like this. Users should be allowed to bring their own event loop and open more than one framebuffer window at once. ;)

LoganDark committed:
Fix examples and doctests and things

LoganDark committed:
Version bump, some doc changes, removal of is_running()

LoganDark committed:
impl Clone for S: Clone

LoganDark committed:
MULTI WINDOW EXAMPLE

LoganDark committed:
minor fix

LoganDark commented:
With a little bit of work, it's now possible to run multiple `mini_gl_fb` windows on the same event loop. I've just committed an example which showcases this. I'd say that this PR is ready for review now. (although custom shader is still broken, game of life works fine, custom shader is probably just old)

LoganDark committed:
properly clamp things in multi_window example

LoganDark committed:
Why not third window? :)

LoganDark commented:
![multi window](https://user-images.githubusercontent.com/4723091/104860822-daff9180-58e1-11eb-885b-6a64c0ff1701.png)

LoganDark committed:
Add an option to turn off Y inversion

LoganDark committed:
Add more doc comments to multi_window

LoganDark committed:
preserve triangle normals? I have no idea what I am doing

LoganDark committed:
Add escape to exit window in multi_window

LoganDark committed:
bresenham line drawing algorithm

LoganDark committed:
Fullscreen support for multi window, because why not

LoganDark committed:
(multi_window) switch to quitting on escape press instead of release

dxenonb commented:
Hey thanks for this! I will look over this PR... "soon." (my timeline is probably measured in weeks 😫) Feel free to ping me in the rust discord if you need more action on it.

I am definitely interested in getting this lib updated!

LoganDark committed:
handle non-inverted y in glutin_handle_basic_input

LoganDark committed:
correct documentation for Config.resizable

LoganDark committed:
WHOOPS remove outdated comment

LoganDark committed:
remove redundant check in multi_window

LoganDark committed:
(multi_window) make the pixels biggerer

LoganDark committed:
(multi_window) rustfmt was only right about one thing

LoganDark committed:
(multi_window) fix cursor accuracy

LoganDark committed:
(multi_window) re-fix cursor accuracy

LoganDark commented:
Screenshots gallery:

![Windows](https://user-images.githubusercontent.com/4723091/104958667-4babba00-5985-11eb-99ce-bce91ece974a.png)
![Arch Linux](https://user-images.githubusercontent.com/4723091/104958676-50706e00-5985-11eb-8fe7-fa0ac0ea310b.png)
![macOS](https://user-images.githubusercontent.com/4723091/104958692-55352200-5985-11eb-9fe2-a1859e6ef478.png)

dxenonb commented on src/config.rs:
```diff
@@ -15,12 +15,27 @@
+    pub invert_y: bool
+}
+
+impl<S: ToString + Clone> Clone for Config<S> {
```
Hmm is the ToString constraint breaking `derive(Clone)`?

Or was this at one point necessary while upgrading but no longer so? (Will need to double check my understanding of Clone.) If `ToString` is an issue it can be removed from the definition of `Config`, but left as a constraint elsewhere where Config is consumed. (Deriving seems nicer to cut down on boilerplate but this feels very nit picky!)

LoganDark replied:
> Hmm is the ToString constraint breaking derive(Clone)?

Not quite, the generic argument itself is, because S does not necessarily implement Clone. :P 

> If ToString is an issue it can be removed from the definition of Config, but left as a constraint elsewhere where Config is consumed.

ToString should remain a constraint so that people can't set the window title to a File handle. Even if no functions will accept that...

dxenonb commented on src/config.rs:
```diff
@@ -31,6 +46,7 @@ impl<'a> Default for Config<&'a str> {
             // :^)
             window_title: "Super Mini GL Framebufferer 3!",
             window_size: (600.0, 480.0),
+            invert_y: true
```
I would probably have more credibility to argue this if I didn't hard cod a silly window title, but invert_y: true feels very opinionated for `Default` 😄

I'm thinking it would be cool to add a parameter somewhere that takes an `Origin::{top/bottom left/right}` enum. Default stops making sense then but could be replaced by `Config::basic_top_left`, or with something to that affect.

LoganDark replied:
> I would probably have more credibility to argue this if I didn't hard cod a silly window title, but invert_y: true feels very opinionated for Default 😄

I'd say it's actually more opinionated on my part, since I'm saying that the buffer starting at the bottom of the window is inverted from the way it should be. :P 

Most people expect libraries that have to do with pixels to have an origin in the top left. The fact that this is the default is just how it used to be. We don't need to make TOO many breaking changes now, do we? :P

dxenonb commented on src/core.rs:
```diff
@@ -258,79 +227,92 @@ impl Internal {
                 let y_scale = self.fb.buffer_height as f64 / (self.fb.vp_height as f64);
                 let mouse_pos = (
                     x * x_scale,
                     // use the OpenGL texture coordinate system instead of window coordinates
```
This is more of a :big thonk: on my side than a comment on this PR - did my comment

> // use the OpenGL texture coordinate system instead of window coordinates

have a reasonable basis in anything? Seems like an odd choice to throw in here. Nice if it was configurable maybe?

LoganDark replied:
The "OpenGL texture coordinate system" you refer to is the one where the origin is at the bottom left. You inverted the mouse Y so that programs that want the mouse position will get the one that's over the pixel at the specified coordinates - one with the origin at the bottom left.

> Nice if it was configurable maybe?

It is, via the `invert_y` option.

dxenonb replied:
I think invert_y is good for now. What I meant was maybe take `origin: Origin` in config, which would let you put the origin in any corner or the center, or something like that.

LoganDark replied:
That would be even more cursed. I think this is fine. The only reason inverted Y exists is because existing applications depend on it... then again, we could just remove it altogether and force everyone to move, as I have with all my breaking changes.

dxenonb commented on src/core.rs:
```diff
@@ -258,79 +227,92 @@ impl Internal {
+
+        event_loop.run_return(|event, _, flow| {
+            // for compatibility with previous versions, which used an infinite run loop
+            *flow = ControlFlow::Poll;
```
Polling seems like the "safe default," (e.g. the user is doing something time based) but I'm open if we want to expose a `Poll` vs `Wait` argument somewhere/somehow

LoganDark replied:
Done

dxenonb commented on examples/multi_window.rs:
```diff
@@ -0,0 +1,282 @@
+                *flow = ControlFlow::Wait;
+
+                self.windows.retain(|window|
+                    unsafe { &mut *window.as_ptr() }.handle_event(&event)
```
👀 but this is not in the core lib so /shrug if it runs 😄 - maybe at least comment it so nobody gets led astray

LoganDark replied:
Good now?

LoganDark committed:
(multi_window) use UnsafeCell instead of Cell

It's better because at least I'm up-front about the fact that I'm totally bypassing the borrow checker here

dxenonb reviewed (commented):
My disposition is merge, since we really need to get winit out of the pre .20 stone age

Only major thoughts are:

 - Can/should we put an events loop reference into the framebuffer struct? Is that gross? I don't love passing it into the input helper methods. Willing to merge for 0.8 though even if we want to fix that later, per stone-age-winit comment.
 - Should fix the unsafe in the multi window example because I'm getty spooky UB vibes from it. Don't want people copying that blindly

LoganDark commented:
> * Can/should we put an events loop reference into the framebuffer struct? Is that gross? I don't love passing it into the input helper methods. Willing to merge for 0.8 though even if we want to fix that later, per stone-age-winit comment.

It's possible to bind the lifetime of the Framebuffer struct to the lifetime of the event loop, if that's what you mean. (Another breaking change, of course.) Otherwise, there is a reason I ripped the event loop out of the library internals - I want to bring my own, I don't want you to own my event loop.

> * Should fix the unsafe in the multi window example because I'm getty spooky UB vibes from it. Don't want people copying that blindly

I'd have to reimplement drain_filter, which is currently an unstable, nightly-only feature. :(

LoganDark committed:
(multi_window) add explanatory comment about UB

LoganDark committed:
Add screenies to README, update copy

LoganDark committed:
Update README.md

LoganDark committed:
Update README.md

LoganDark committed:
Update README.md

LoganDark committed:
FramebufferInternal

LoganDark committed:
GlutinBreakout::make_current

LoganDark committed:
Documentation

LoganDark committed:
(multi_window) remove the unsafe mess

LoganDark committed:
Make GlutinBreakout::make_current look a lot nicer

LoganDark committed:
docs

LoganDark committed:
docs

LoganDark committed:
fix compilation

LoganDark committed:
(multi_window) make resizing a bit more efficient

dxenonb commented on README.md:
```diff
@@ -4,23 +4,43 @@
 
 Mini GL Framebuffer provides an easy way to draw to a window from a pixel buffer. OpenGL
-alternative to other easy framebuffer libraries.
+alternative to other easy framebuffer libraries like `minifb` and `pixels`.
```
Note to self, add links later to be a cool cat in the ecosystem

Should we spell out differences more too later? (Not necessary for this PR to be merged)

LoganDark replied:
> Should we spell out differences more too later?

Could add a feature matrix in the docs, if you like.

dxenonb commented on README.md:
```diff
@@ -4,23 +4,43 @@
+OpenGL.
+
+# Screenies
+Here are some screenies of the `multi_window` example running on different platforms:
```
`Screenies` might be too fun for me 😸

LoganDark replied:
Super Mini GL Framebufferer 3

LoganDark committed:
Remove generic argument from Config; add docs; use LogicalSizes

LoganDark committed:
Allow basic_input consumers to use ControlFlow::Wait

LoganDark committed:
Derives

LoganDark committed:
make Config non_exhaustive; add config! macro

LoganDark committed:
add comment in gotta_go_fast about why it doesn't use the config! macro

LoganDark committed:
Make BasicInput derive Debug

LoganDark committed:
Link to MiniGlFb::internal

LoganDark committed:
improve documentation intralinking; improve docs especially around the config macro

LoganDark committed:
add spread support to config! macro

LoganDark committed:
doc fix

LoganDark committed:
Use $crate in `config!`

LoganDark committed:
slight grammar fix

LoganDark committed:
Add feature matrix

LoganDark committed:
Check for Esc press instead of release

LoganDark committed:
Add non_exhaustive to Framebuffer struct

dxenonb commented:
Awesome stuff. Great to get glutin upgraded. Thanks. Merging!
